functorch/experimental/_cond.py

```diff
@@ -193,18 +200,41 @@ def _has_potential_branch_input_alias(branch, inputs):
     except Exception as e:
         raise e
 
-    input_storages = set()
-    for node in gm.graph.nodes:
-        if node.op == "placeholder":
-            input_storages.add(StorageWeakRef(node.meta['val']._typed_storage()))
-        if node.op == "output":
-            for out in node.args:
-                out_storage = StorageWeakRef(out.meta["val"]._typed_storage())
-                if out_storage in input_storages:
-                    return True
+    def _detect_input_alias(gm):
+        input_storages = set()
+        for node in gm.graph.nodes:
+            if node.op == "placeholder":
+                input_storages.add(StorageWeakRef(node.meta['val']._typed_storage()))
+            if node.op == "output":
+                for out in node.args:
+                    out_storage = StorageWeakRef(out.meta["val"]._typed_storage())
+                    if out_storage in input_storages:
+                        return True
+
+        for _, module in gm.named_children():
+            if isinstance(module, torch.fx.GraphModule) and _detect_input_alias(module):
+                return True
+
+        return False
+
+    return _detect_input_alias(gm)
+
+
+@cond.py_impl(DispatchKey.Functionalize)
+def cond_func(pred, true_fn, false_fn, inputs):
+    unwrapped_inputs = _unwrap_all_tensors_from_functional(inputs, reapply_views=torch._C._functionalization_reapply_views_tls())
+    unwrapped_pred = _unwrap_all_tensors_from_functional(pred, reapply_views=torch._C._functionalization_reapply_views_tls())
+    guard = ExcludeDispatchKeyGuard(DispatchKeySet(DispatchKey.Functionalize))
+    for branch in [true_fn, false_fn]:
+        if _has_potential_branch_input_mutation(branch, unwrapped_inputs):
+            raise UnsupportedAliasMutationException("One of torch.cond branch "
+                                                    "might be modifying the input!")
 
-    return False
+        if _has_potential_branch_input_alias(branch, unwrapped_inputs):
+            raise UnsupportedAliasMutationException("One of torch.cond branch "
+                                                    "might be aliasing the input!")
 
+    return _wrap_all_tensors_to_functional(cond(unwrapped_pred, true_fn, false_fn, unwrapped_inputs), level=0)
```

Comment:
I don't think this actually works -- we disabled the functionalization key, so the branches will not get run through with functional tensors. I suggested some test cases below, it doesn't look like we are actually testing any conds with in-place ops in the branch functions.

My take is we still need to apply functionalize(true_fn) and functionalize(false_fn) here